java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/buildsteps/RuntimeImageBuildStep.java

```diff
@@ -24,41 +24,78 @@
 import com.google.cloud.runtimes.builder.config.domain.JdkServerLookup;
 import com.google.cloud.runtimes.builder.config.domain.RuntimeConfig;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
 public abstract class RuntimeImageBuildStep implements BuildStep {
 
+  private final Logger logger = LoggerFactory.getLogger(RuntimeImageBuildStep.class);
+
   private final JdkServerLookup jdkServerLookup;
+  private final String compatImageName;
 
-  protected RuntimeImageBuildStep(JdkServerLookup jdkServerLookup) {
+  protected RuntimeImageBuildStep(JdkServerLookup jdkServerLookup, String compatImageName) {
     this.jdkServerLookup = jdkServerLookup;
+    this.compatImageName = compatImageName;
   }
 
   @Override
   public void run(BuildContext buildContext) throws BuildStepException {
-    buildContext.getDockerfile().appendLine("FROM " + getBaseRuntimeImage(buildContext));
+    Path artifact = getArtifact(buildContext);
+    logger.debug("Found Java artifact {}", artifact);
+
+    String baseImage;
+    try {
+      baseImage = getBaseRuntimeImage(buildContext, artifact);
+    } catch (IOException e) {
+      throw new BuildStepException("An error was encountered while searching for an artifact. "
+          + "Please try again later.", e);
+    }
 
+    buildContext.getDockerfile().appendLine("FROM " + baseImage);
     String copyStep = "COPY ";
     if (buildContext.isSourceBuild()) {
       copyStep += "--from=" + DOCKERFILE_BUILD_STAGE + " ";
     }
-    buildContext.getDockerfile().appendLine(copyStep + getArtifact(buildContext)
-        + " $APP_DESTINATION");
+
+    String relativeArtifactPath = "./" + buildContext.getWorkspaceDir().relativize(artifact)
+        .toString();
+    buildContext.getDockerfile().appendLine(copyStep + relativeArtifactPath + " $APP_DESTINATION");
```

Comment:
Compat image doesn't define `$APP_DESTINATION`. So, I think we should just hardcode the destination path for compat.